src/persist-client/src/bin/persist_open_loop_benchmark.rs

```diff
@@ -0,0 +1,449 @@
+            // as a relative duration from start.
+            let mut prev_log = Duration::from_millis(0);
+            let mut batches_written = 0;
+
+            loop {
+                let batch = match rx.recv().await {
+                    Some(batch) => batch,
+                    None => {
+                        let elapsed = start.elapsed();
+                        let records_written = batches_written * args.batch_size;
+                        let finished = format!("Writer {} finished after {} ms and wrote {} records. Max write latency {} ms.",
+                          idx, elapsed.as_millis(), records_written, max_write_latency.as_millis());
+                        return Ok(finished);
+                    }
+                };
+
+                trace!("writer {} received a batch. writing", idx);
+                let write_start = Instant::now();
+                writer.write(batch).await?;
+                batches_written += 1;
+                let write_latency = write_start.elapsed();
+                if write_latency > max_write_latency {
+                    max_write_latency = write_latency;
+                }
+
+                let elapsed = start.elapsed();
+
+                if elapsed - prev_log > FIVE_SECONDS {
+                    let records_written = batches_written * args.batch_size;
+                    info!("After {} ms writer {} has written {} records. Max write latency {} ms most recent write latency {} ms.",
+                          elapsed.as_millis(), idx, records_written, max_write_latency.as_millis(), write_latency.as_millis());
+                    prev_log = elapsed;
+                }
+            }
+        });
+
+        handles.push(writer_handle);
+    }
+
+    for (idx, mut reader) in readers.into_iter().enumerate() {
+        let b = Arc::clone(&barrier);
+        let reader_handle = mz_ore::task::spawn(|| format!("reader-{}", idx), async move {
+            trace!("reader {} waiting for barrier", idx);
+            b.wait().await;
+            info!("starting reader {}", idx);
+
+            // Max observed latency for BenchmarkReader::num_records.
+            let mut max_read_latency = Duration::from_millis(0);
+
+            // Max observed delay between the number of records expected to be read at any
+            // point in time vs the number of records actually ingested by that point.
+            let mut max_lag = 0;
+
+            // The last time we emitted progress information to stdout, expressed
+            // as a relative duration from start.
+            let mut prev_log = Duration::from_millis(0);
+            loop {
+                let elapsed = start.elapsed();
+                let expected_sent =
+                    elapsed.as_millis() as usize / (millis_per_batch as usize) * args.batch_size;
+                let read_start = Instant::now();
+                let num_records_read = reader.num_records().await?;
```

Comment:
This call only polls the next batch of records and then returns. I think we need to spawn a reader task that continuously polls records and updates an atomic that allows getting the current number of records read.

Reply:
Oh interesting -- I guess the analogue for the old test would be:

in the test we were dumping data into kafka on a schedule. Interally, materialize was continually reading from that topic and computing how many records it had read, and populating that in a materialized view, and for the test, we were querying that materialized view with a fast query to see how many records mz had read.

In this version -- the separate task that is continuously reading and updating an atomic is analogous to Materialize reading from the source and populating a materialized view, and then reading that atomic is analogous to querying that materialized view.

I'll make this change right away.

Reply:
Aljoscha and I chatted in person. The takeaway was:

We don't actually need to spawn a different task to read continuously because - this reader task is already reading continuously and everything else its doing is computationally minimal. Having a separate task that _only_ polled could make that faster, but then it would introduce updating an atomic and a new task to read the atomic which could make the benchmark overhead worse.

Conclusion - not going to change anything for now.